### PR TITLE
Expose viewer controls to pipeline

### PR DESCRIPTION
**Pull request by kerrj**
PR to track exposing viewer controls to modules in the pipeline. Not finished, but crop params are implemented as an example of how we might extend this to other capabilities like camera movement (see ViewerControl inside viewer_elements.py)

### CODE REVIEW

File: nerfstudio/viewer/viser/messages.py
Comment (tancik) [inline nerfstudio/viewer/viser/messages.py:1]: You should pull main and rerun the sync command. It will now update the package version

File: docs/developer_guides/viewer/viewer_control.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Python Viewer Control
+
+Similar to `ViewerElements`, Nerfstudio includes supports a pythonic interface to the viewer through which you can:
```
Comment (tancik) [inline docs/developer_guides/viewer/viewer_control.md:3]: `ViewerElements` -> `[ViewerElements](./custom_gui.md)`

File: docs/developer_guides/viewer/viewer_control.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            #example of using the get_camera function, pass img width and height
+            #returns a Cameras object with 1 camera
+            camera = self.viewer_control.get_camera(100,100)
+            if camera is None:
+                #returns None when the viewer is not connected yet
+                return
+            #get the camera pose
+            camera_extrinsics_matrix = camera.camera_to_worlds[0,...]#3x4 matrix
+            #generate image RayBundle
+            bundle = camera.generate_rays(camera_indices=0)
```
Comment (tancik) [inline docs/developer_guides/viewer/viewer_control.md:43]:
```
bundle = camera.generate_rays(camera_indices=0)
# Compute loss, or whatever you want
...
```

File: docs/developer_guides/viewer/viewer_control.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+Just like `ViewerElements`, you can create an instance inside any class which inherits from `nn.Module`
+and is contained within the `Pipeline` object (for example the `Model`)
+
+```python
+from nerfstudio.viewer.server.viewer_elements import ViewerControl
+
+class MyModel(nn.Module):#must inherit from nn.Module
+    def __init__(self):
+        # Must be a class variable
+        self.viewer_control = ViewerControl()#no arguments
```
Comment (tancik) [inline docs/developer_guides/viewer/viewer_control.md:22]: Can you use the same spacing that black enforces for comments, ie 
`ViewerControl()..#.comment`
Here and elsewhere

File: nerfstudio/viewer/app/src/modules/Scene/Scene.jsx [hunk 4]
```diff
@@ -107,20 +111,20 @@ export function get_scene_tree() {
       camera_controls.setTarget(
         curPos.x + diff.x,
         curPos.y + diff.y,
-        curPos.z + diff.z,
+        curPos.z + diff.z,true
```
Comment (tancik) [inline nerfstudio/viewer/app/src/modules/Scene/Scene.jsx:114]: Write out argument name, hard to tell what this is referring to. (here and below)

Comment (tancik) [inline nerfstudio/viewer/app/src/modules/Scene/Scene.jsx:114]: Run `prettier`, it doesn't look like any formatting rules are being applied.

File: nerfstudio/viewer/app/src/modules/ViewerWindow/ViewerWindow.jsx [hunk 5]
```diff
@@ -313,6 +313,49 @@ export default function ViewerWindow(props) {
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [isWebsocketConnected]);
 
+  const throttledClickSender = makeThrottledMessageSender(
+    viser_websocket,
+    10,
+  );
+  useEffect(() => {
+    const onMouseDouble = (e) => {
+      const BANNER_HEIGHT = 50;
```
Comment (tancik) [inline nerfstudio/viewer/app/src/modules/ViewerWindow/ViewerWindow.jsx:322]: Not sure how, but this should be a same value here and in `scene.jsx`. Could cause bugs in the future if one is updated but not the other.

File: nerfstudio/viewer/app/src/modules/WebSocket/ViserWebSocket.tsx [hunk 6]
```diff
@@ -159,6 +162,37 @@ function handleMessage(
       }
       break;
     }
+    // Set camera position
+    case 'SetCameraMessage': {
+      if(message.fov !== null){
+        // sceneTree.current.metadata.camera.fov = message.fov;
```
Comment (tancik) [inline nerfstudio/viewer/app/src/modules/WebSocket/ViserWebSocket.tsx:168]: Delete

File: nerfstudio/viewer/app/src/modules/WebSocket/ViserWebSocket.tsx [hunk 7]
```diff
@@ -291,7 +325,9 @@ function handleMessage(
   }
 }
 
-export function ViserWebSocket({ children }: { children: React.ReactNode }) {
+// export function ViserWebSocket({ children }: { children: React.ReactNode }) {
+export function ViserWebSocket({sceneTree, children}: {sceneTree: SceneNode, children: React.ReactNode}) {
+  // const sceneTree = props.sceneTree;
```
Comment (tancik) [inline nerfstudio/viewer/app/src/modules/WebSocket/ViserWebSocket.tsx:330]: delete

File: nerfstudio/viewer/server/viewer_elements.py [hunk 8]
```diff
@@ -16,18 +16,130 @@
...
 
+@dataclass
+class ViewerClick:
+    """
+    Class representing a click in the viewer as a ray. Implemented as a class to support future new information possibly
+    """
+
+    # the information here matches the information in the ClickMessage,
+    # but we implement a wrapper as an abstraction layer
+    origin: Tuple[float, float, float]
```
Comment (tancik) [inline nerfstudio/viewer/server/viewer_elements.py:48]: Add docstrings

File: nerfstudio/viewer/server/viewer_elements.py [hunk 9]
```diff
@@ -16,18 +16,130 @@
...
+    from nerfstudio.viewer.server.viewer_state import ViewerState
+
 
 TValue = TypeVar("TValue")
 
 
+@dataclass
+class ViewerClick:
+    """
+    Class representing a click in the viewer as a ray. Implemented as a class to support future new information possibly
```
Comment (tancik) [inline nerfstudio/viewer/server/viewer_elements.py:43]: Delete, "Implemented as a class to support future new information possibly"

File: nerfstudio/viewer/viser/message_api.py [hunk 10]
```diff
@@ -448,6 +448,20 @@ def set_training_state(self, training_state: Literal["training", "paused", "comp
...
+        self,
+        position: Optional[Tuple[float, float, float]] = None,
+        look_at: Optional[Tuple[float, float, float]] = None,
+        fov: Optional[int] = None,
+        instant: bool = False,
+    ) -> None:
+        """Update the camera object in the viewer.
+
+        Args:
+            camera: The new camera
```
Comment (tancik) [inline nerfstudio/viewer/viser/message_api.py:461]: Update args

File: nerfstudio/viewer/viser/messages.py [hunk 11]
```diff
@@ -243,6 +256,14 @@ class TimeConditionMessage(NerfstudioMessage):
     """ Time conditioning value """
 
 
+@dataclasses.dataclass
+class ClickMessage(NerfstudioMessage):
+    """Click message."""
+
+    origin: Tuple[float, float, float]
+    direction: Tuple[float, float, float]
```
Comment (tancik) [inline nerfstudio/viewer/viser/messages.py:264]: Add docstrings

**Review by tancik — COMMENTED** [review]
This will be super handy in future methods. Thanks for adding.

File: docs/developer_guides/viewer/viewer_control.md (hunk 1 above)
Comment (brentyi) [inline docs/developer_guides/viewer/viewer_control.md:3]: pythonic -> Pythonic?

File: docs/developer_guides/viewer/viewer_control.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+First, instantiate a `ViewerControl` object as a class variable inside a model file.
+Just like `ViewerElements`, you can create an instance inside any class which inherits from `nn.Module`
+and is contained within the `Pipeline` object (for example the `Model`)
+
+```python
+from nerfstudio.viewer.server.viewer_elements import ViewerControl
+
+class MyModel(nn.Module):#must inherit from nn.Module
+    def __init__(self):
+        # Must be a class variable
```
Comment (brentyi) [inline docs/developer_guides/viewer/viewer_control.md:21]: capitalization could be made consistent (personally I prefer capitalized, but not sure what repo mostly has)

File: docs/developer_guides/viewer/viewer_control.md [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        self.viewer_control = ViewerControl()#no arguments
+        def click_cb(click:ViewerClick):
+            print(f"Click at {click.origin} in direction {click.direction}")
+        self.viewer_control.register_click_cb(click_cb)
+```
+
+### Thread safety
+Keep in mind that, just like `ViewerElement` callbacks, click callbacks are asynchronous to training and can potentially interrupt a call to get_outputs.
```
Comment (brentyi) [inline docs/developer_guides/viewer/viewer_control.md:79]: `Keep in mind that, just like` => `Just like`, get_outputs => `get_outputs()`

File: nerfstudio/scripts/viewer/sync_viser_message_defs.py [hunk 14]
```diff
@@ -34,7 +34,7 @@ def main() -> None:
     # - nerfstudio/scripts/viewer/
     # - nerfstudio/scripts/
     # - nerfstudio/
-    target_path = pathlib.Path(__file__).absolute().parent.parent.parent / pathlib.Path(
+    target_path = pathlib.Path(__file__).absolute().parent.parent.parent.parent / pathlib.Path(
```
Comment (brentyi) [inline nerfstudio/scripts/viewer/sync_viser_message_defs.py:37]: explain? (If this is updated it seems like the comment above should be as well?)

File: nerfstudio/viewer/app/src/modules/ViewerWindow/ViewerWindow.jsx [hunk 15]
```diff
@@ -191,7 +191,7 @@ export default function ViewerWindow(props) {
         const oldBackground = scene.background;
         const texture = new THREE.Texture(this);
         texture.minFilter = THREE.LinearFilter;
-        texture.magFilter = THREE.NearestFilter;
+        texture.magFilter = THREE.LinearFilter;
```
Comment (brentyi) [inline nerfstudio/viewer/app/src/modules/ViewerWindow/ViewerWindow.jsx:194]: don't feel strongly but would be cleaner to move changes that are orthogonal to viewercontrols to their own PRs

File: nerfstudio/viewer/app/src/modules/ViewerWindow/ViewerWindow.jsx (hunk 5 above)
Comment (brentyi) [reply nerfstudio/viewer/app/src/modules/ViewerWindow/ViewerWindow.jsx:322]: maybe `myRef.current.offsetTop` ? we should probably also name this something more meaningful

File: nerfstudio/viewer/viser/messages.py [hunk 16]
```diff
@@ -124,6 +124,19 @@ class FilePathInfoMessage(NerfstudioMessage):
...
+class SetCameraMessage(NerfstudioMessage):
+    """Set the current camera."""
+
+    fov: Optional[float]
+    """ Field of view of the camera """
+    look_at: Optional[Tuple[float, float, float]]
+    """Point in 3D the camera is looking at"""
+    position: Optional[Tuple[float, float, float]]
+    """ Position of the camera"""
+    instant: bool = False
```
Comment (brentyi) [inline nerfstudio/viewer/viser/messages.py:137]: probably don't need this default value, also missing a docstring?

File: nerfstudio/viewer/viser/messages.py [hunk 17]
```diff
@@ -124,6 +124,19 @@ class FilePathInfoMessage(NerfstudioMessage):
     """ Name of the export folder """
 
 
+@dataclasses.dataclass
+class SetCameraMessage(NerfstudioMessage):
+    """Set the current camera."""
+
+    fov: Optional[float]
+    """ Field of view of the camera """
```
Comment (brentyi) [inline nerfstudio/viewer/viser/messages.py:132]: would be nice to make the spaces/punctuation consistent etc in the docstring

File: nerfstudio/scripts/viewer/sync_viser_message_defs.py (hunk 14 above)
Comment (kerrj) [reply nerfstudio/scripts/viewer/sync_viser_message_defs.py:37]: @tancik and I changed this after ns-dev-sync failed because the path of the scripts/ directory changed

Comment (tancik) [reply nerfstudio/scripts/viewer/sync_viser_message_defs.py:37]: The comment still holds, I think it was wrong originally.

File: nerfstudio/viewer/app/src/modules/ViewerWindow/ViewerWindow.jsx (hunk 5 above)
Comment (chungmin99) [reply nerfstudio/viewer/app/src/modules/ViewerWindow/ViewerWindow.jsx:322]: I agree that `BANNER_HEIGHT` should be synced between the two files.
Seems like `myRef.current.offsetTop` returns 0 right now though (instead of expected 50), this might need some more investigation...

File: nerfstudio/scripts/viewer/sync_viser_message_defs.py (hunk 14 above)
Comment (brentyi) [reply nerfstudio/scripts/viewer/sync_viser_message_defs.py:37]: It looks to me like we should drop one `.parent`, and remove the `nerfstudio/` in the string on the next line

Comment (kerrj) [reply nerfstudio/scripts/viewer/sync_viser_message_defs.py]: fixed

**Review by tancik — APPROVED** [review]
LGTM
